### PR TITLE
[go_router] Add documentation to some methods in matching.dart

### PR DESCRIPTION
**Pull request by ValentinVignal**
I was investigating how to fix https://github.com/flutter/flutter/issues/113032 (I will soon public a design proposition as I think the entire `RouteMatchList` structure needs to be changed from a list to a tree). It is a bit hard to understand what each method was supposed to do so I added some documentation to help future contributors.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/go_router/lib/src/matching.dart [hunk 1]
```diff
@@ -76,8 +83,11 @@ class RouteMatchList {
 
   final List<RouteMatch> _matches;
 
-  /// the full path pattern that matches the uri.
-  /// /family/:fid/person/:pid
+  /// the full path pattern that matches the uri. For example:
```
Comment (chunhtai) [inline packages/go_router/lib/src/matching.dart:86]:
```suggestion
  /// The full path pattern that matches the uri.
  ///
  /// For example:
```

File: packages/go_router/lib/src/matching.dart [hunk 2]
```diff
@@ -159,6 +169,17 @@ class MatcherError extends Error {
   }
 }
 
+/// Returns the list of `RouteMatch` corresponding to the given [loc]. For
+/// example, for a given [loc] `/a/b/c/d`, this function will return the list of
+/// [RouteBase] `[GoRouteA(), GoRouterB(), GoRouteC(), GoRouterD()]`.
```
Comment (chunhtai) [inline packages/go_router/lib/src/matching.dart:174]:
```suggestion
/// Returns the list of `RouteMatch` corresponding to the given [loc].
///
/// For example, for a given [loc] `/a/b/c/d`, this function will return the list of
/// [RouteBase] `[GoRouteA(), GoRouterB(), GoRouteC(), GoRouterD()]`.
```

File: packages/go_router/lib/src/matching.dart [hunk 3]
```diff
@@ -58,6 +60,11 @@ class RouteMatchList {
   static RouteMatchList empty =
       RouteMatchList(<RouteMatch>[], Uri.parse(''), const <String, String>{});
 
+  /// Generate the full path (ex: `'/family/:fid/person/:pid'`) of a list of
+  /// [RouteMatch].
+  ///
+  /// This methods considers that [matches] is ordered accorresponding to how
```
Comment (chunhtai) [inline packages/go_router/lib/src/matching.dart:66]: I can't quite parse this sentence

File: packages/go_router/lib/src/matching.dart [hunk 4]
```diff
@@ -159,6 +169,17 @@ class MatcherError extends Error {
   }
 }
 
+/// Returns the list of `RouteMatch` corresponding to the given [loc]. For
+/// example, for a given [loc] `/a/b/c/d`, this function will return the list of
+/// [RouteBase] `[GoRouteA(), GoRouterB(), GoRouteC(), GoRouterD()]`.
+///
+/// - [loc] is the complete URL to match (without the query parameters). For
+///   example, for the URL `/a/b?c=0`, [loc] will be `/a/b`.
+/// - [restBloc] is the remaining part of the URL to match while [parentBloc] is
```
Comment (chunhtai) [inline packages/go_router/lib/src/matching.dart:178]: restLoc and parenetLoc?

File: packages/go_router/lib/src/matching.dart (hunk 3 above)
Comment (ValentinVignal) [reply packages/go_router/lib/src/matching.dart]: I hope [docs: Better documentation](https://github.com/flutter/packages/pull/3462/commits/f12243ac360b25394893e7e0b4cbf5352ebd45be) makes it clearer

**Comment by ValentinVignal** [comment]
> I would think this is a version exempt ?

Should the label "override: no versioning needed" be added to this PR ? Or should I create a new version?

File: packages/go_router/lib/src/matching.dart [hunk 5]
```diff
@@ -58,6 +60,34 @@ class RouteMatchList {
   static RouteMatchList empty =
       RouteMatchList(<RouteMatch>[], Uri.parse(''), const <String, String>{});
 
+  /// Generate the full path (ex: `'/family/:fid/person/:pid'`) of a list of
```
Comment (chunhtai) [inline packages/go_router/lib/src/matching.dart:63]:
```suggestion
  /// Generates the full path (ex: `'/family/:fid/person/:pid'`) of a list of
```

File: packages/go_router/lib/src/matching.dart [hunk 6]
```diff
@@ -58,6 +60,34 @@ class RouteMatchList {
...
+  /// [RouteMatch].
+  ///
+  /// This methods considers that [matches]'s elements verify the go route
```
Comment (chunhtai) [inline packages/go_router/lib/src/matching.dart:66]: Maybe:
```
The [matches] must be the in same order of how GoRoutes are matched.

For example, if the routes structure is ...
```

**Review by chunhtai — APPROVED** [review]
LGTM, just some more nits

**Review by johnpryan — APPROVED** [review]

**Pull request merged**